sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveTooManyLines.cs

```diff
@@ -0,0 +1,117 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class MethodsShouldNotHaveTooManyLines : ParameterLoadingDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S138";
+        private const string MessageFormat = "This {0} has {1} lines, which is greater than the {2} lines authorized. Split it into smaller methods.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private const int DefaultMaxMethodLines = 80;
+        [RuleParameter("max", PropertyType.Integer, "Maximum authorized lines of code in a method",
+            DefaultMaxMethodLines)]
+
+        public int Max { get; set; } = DefaultMaxMethodLines;
+        
+        protected override void Initialize(ParameterLoadingAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    if (Max < 2)
+                    {
+                        throw new InvalidOperationException($"Invalid maximum number of lines: {Max}");
+                    }
+
+                    if (c.IsTest())
+                    {
+                        return;
+                    }
+
+                    var baseMethodSyntax = (BaseMethodDeclarationSyntax)c.Node;
+
+                    var identifierLocation = baseMethodSyntax?.FindIdentifierLocation();
+                    var bodyStatements = baseMethodSyntax?.Body?.Statements;
+                    if (bodyStatements == null || identifierLocation == null)
+                    {
+                        return;
+                    }
+
+                    var linesCount = bodyStatements.Value
+                        .SelectMany(s => s.DescendantTokens())
+                        .Select(token => token.GetLineNumber())
+                        .Distinct()
+                        .LongCount();
+
+                    if (linesCount > Max)
+                    {
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, identifierLocation,
+                            GetDescription(baseMethodSyntax), linesCount, Max));
+                    }
+                },
+                SyntaxKind.MethodDeclaration,
+                SyntaxKind.ConstructorDeclaration,
+                SyntaxKind.DestructorDeclaration);
```

Comment:
Do we care about properties? I am sure that you will find positives in the ITs :)